Product Collection: Add `Attributes` filter control to sidebar settings

imanish003 opened the pull request:
This PR adds `Attributes` filter control to sidebar settings.
<img width="266" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/a7aac444-d01d-40c1-8e0d-66883d083e4c">

Fixes #9363 

### Testing

1. Create a new post
2. Add Product Collection block
3. With the `Product Collection` block selected, inspect the sidebar for the `Product attributes` control. Confirm its appearance and functionalities:
   - The control should be displayed as shown in the images below:
<img width="280" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/459ffec4-5276-4a95-a3ca-3e4f6762bea9">
<img width="266" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/a7aac444-d01d-40c1-8e0d-66883d083e4c">

4. Test the `Product attributes` control by selecting any attributes. For example, color -> blue:
   - After the selection, verify that changes are immediately visible in the Editor view.

5. Save and publish the post:
   - Click on `Publish...` at the top right corner.
   - After publishing, navigate to the Frontend and verify that the changes you made in the Editor are accurately reflected.

6. Return to the Editor and validate the `Reset all` functionality of the filters:
   - Make sure that clicking the `Reset all` button resets your selected filters.
   - The `Reset all` button can be found as shown in the image below:
<img width="270" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/fa14b1c9-3db9-4743-af97-a543b7e9a90f">

7. Test the `Manage attributes` functionality:
   - Click on the `Manage attributes` link.
   - Verify that this redirects you to the designated page where you can manage product attributes


* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

### Changelog

> Product Collection: Add `Attributes` filter control to sidebar settings

imanish003 committed:
Add columns control to product collection block editor settings

- `InspectorControls` from './inspector-controls' is now imported in `edit.tsx` and used in the returned JSX of `Edit` function.
- A new file `columns-control.tsx` is added under 'product-collection' block's 'inspector-controls' directory which exports a `ColumnsControl` component. This component uses `RangeControl` from '@wordpress/components' to control the number of columns in the product collection display layout when the layout type is 'flex'.
- The types file (`types.ts`) for 'product-collection' block is updated. The `Attributes` interface is renamed to `ProductCollectionAttributes` and the `ProductCollectionContext` interface is removed. The `ProductCollectionAttributes` now includes 'queryContext', 'templateSlug', and 'displayLayout' properties.

imanish003 committed:
Merge branch 'trunk' into 9356-product-collection-editor-settings-columns

imanish003 committed:
Refactor: Simplify Fallback Return in ColumnsControl Component

This commit simplifies the fallback return value of the ColumnsControl component. Instead of returning an empty fragment (<> </>), it now returns null when the condition isn't met. This change improves readability and aligns with best practices for conditional rendering in React.

imanish003 committed:
Feature: Add 'Order By' Control to Product Collection Inspector

This commit adds a new 'Order By' control to the product collection inspector. The control allows users to specify the order of products in a collection by various attributes such as title and date. To support this, a new component 'OrderByControl' has been created and included in the product collection inspector. Additionally, the types for 'order' and 'orderBy' attributes have been updated and exported for reuse.

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Add more options to OrderBy type

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Add orderby handling on frontend & editor

The main changes include:
1. Added a new property 'isProductCollectionBlock' in the block.json to denote if a block is a product collection block.
2. In the ProductCollection PHP class, a new initialization function has been defined to hook into the WordPress lifecycle, register the block, and update the query based on this block.
3. Added methods to manage query parameters for both frontend rendering and the Editor.
4. Expanded allowed 'collection_params' for the REST API to include custom 'orderby' values.
5. Defined a function to build the query based on block attributes, filters, and global WP_Query.
6. Created utility functions to handle complex query operations such as merging queries, handling custom sort values, and merging arrays recursively.

These improvements allow for more flexible and robust handling of product collections in both the front-end display and the WordPress editor. It also extends support for custom 'orderby' values in the REST API, which allows for more advanced sorting options in product collections.

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Merge branch 'trunk' into 9359-product-collection-editor-settings-order-by

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9361-product-collection-filters-on-sale

imanish003 committed:
Add 'on sale' filter and enhance settings management in product collection block

This commit introduces several changes to the product collection block.
- First, it adds a new 'on sale' filter that can be used to display only the products that are currently on sale.
- It also refactors the settings management in the product collection block to use the experimental ToolsPanel component from WordPress, which provides a more flexible and intuitive way to manage block settings.
- It moves the 'Columns' control into the ToolsPanel, along with the 'Order by' control.
- A new utility function `setQueryAttribute` is introduced to simplify setting nested query parameters.
- The structure of the `ProductCollectionAttributes` and `ProductCollectionQuery` types have been adjusted to accommodate the changes.
- Finally, it makes corresponding changes in the PHP part to handle the new 'on sale' query parameter.

This should enhance the flexibility and user-friendliness of the product collection block.

imanish003 committed:
Merge branch 'trunk' into 9361-product-collection-filters-on-sale

imanish003 committed:
Add stock status filter to WooCommerce product collection block

This commit introduces a stock status filter to the WooCommerce product collection block.

The changes include:
1. Added the ability to filter products based on their stock status within the 'product-collection' block. A new stock status control is created within the inspector-controls of the block.
2. A new 'get_stock_status_query' function is introduced in 'ProductCollection.php' which returns a query for products depending on their stock status.

Please note that the stock status filter will only appear in the experimental build for now.

imanish003 committed:
Refactor Stock Status control of Product Collection block

This commit refactors the Stock Status control. The changes aim to improve the code organization and make the behavior of the component more explicit.

The key modifications are:
1. Moved stock status related constants and functions from `inspector-controls/utils.tsx` to `inspector-controls/constants.ts`. This is done to ensure that all constants and similar utility functions are organized in one place.
2. Updated `product-collection/index.tsx` to import `getDefaultStockStatuses` from `inspector-controls/constants` instead of `inspector-controls/utils`.
3. Updated `stock-status-control.tsx` to determine whether the stock status has value or not by comparing with the default stock statuses using `fastDeepEqual`. If the stock status control is deselected, it resets the stock status to the default statuses.

These changes do not introduce any new functionalities, but improve the readability and maintainability of the code.

imanish003 committed:
Merge branch 'trunk' into 9362-product-collection-filters-stock-status

imanish003 committed:
Add keyword search control to Product Collection block

This commit introduces a keyword search functionality to the Product Collection block. The update is aimed to provide users with more flexibility and precision in product collection queries.

Key changes:
1. Introduced a new file `keyword-control.tsx` that creates a Keyword Control component. This component includes a TextControl field that allows inputting a search keyword. The keyword search is debounced to prevent unnecessary queries during input and updates the block's attributes accordingly.
2. Modified `inspector-controls/index.tsx` to include the KeywordControl in the ToolsPanel for the block's filters.
3. Adjusted `ProductCollection.php` to include the keyword search in the product query array.

With these changes, users can now search for products by keyword in the Product Collection block.

imanish003 committed:
Resolve conflicts

imanish003 committed:
Add product attributes filter control to ProductCollection block

- This commit introduces the ability to filter products by attributes in ProductCollection block.
- A new `woocommerceAttributes` key was added to the `block.json` file and the `ProductCollectionQuery` type. Also, a new file `attributes-control.tsx` was created, providing the UI component for the attribute filter control in the editor.
- In addition, updates were made to the `ProductCollection.php` file in the backend to support filtering products by attributes, and the tax query was updated to include attribute queries.
- Lastly, the `ProductCollectionInspectorControls` was updated to include the `AttributesControl` component, thus enabling users to filter products by attributes in the block editor."`

imanish003 committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

thealexandrelara committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

thealexandrelara commented:
Hey @imanish003,
The Product Attributes filter is not being displayed for me, is there any requirement or step that I should do to enable it? 

![image](https://github.com/woocommerce/woocommerce-blocks/assets/20469356/319c1e89-b09f-4c15-838b-5b8e77387265)

imanish003 committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

imanish003 commented:
Hey @thealexandrelara,

> The Product Attributes filter is not being displayed for me, is there any requirement or step that I should do to enable it?

That's strange. I can't think of any specific requirement or step to enable it. Are you using a local build? If so, you could try stopping the server and then running `npm start` again.

I've tested it on my end, and it's working fine. Here's a quick demo:

https://github.com/woocommerce/woocommerce-blocks/assets/16707866/aaca8f7c-b8ad-47bd-8ecb-1e097522d638

imanish003 committed:
Resolve conflicts

imanish003 committed:
remove unused import of getDefaultStockStatuses

imanish003 committed:
Resolve conflicts

imanish003 committed:
Delete a duplicate file

imanish003 committed:
Resolve conflicts

thealexandrelara commented:
> That's strange. I can't think of any specific requirement or step to enable it. Are you using a local build? If so, you could try stopping the server and then running npm start again.

Yes, it is a local build. I tried start the server again, disabling Gutenberg, running in Incognito mode but the filter is still not showing to me. Maybe we should ask someone else to also review this PR, so we can confirm if it is only something happening on my end

imanish003 committed:
Remove console log

imanish003 commented:
@thealexandrelara 

> Yes, it is a local build. I tried start the server again, disabling Gutenberg, running in Incognito mode but the filter is still not showing to me. Maybe we should ask someone else to also review this PR, so we can confirm if it is only something happening on my end

Yes, that sounds like a good idea. I am requesting a review from @Aljullu. 🙂 

@Aljullu, FYI, it seems that @thealexandrelara can't see the `Attributes` filter control. Can you please review and check if you are able to see it? 🙂

imanish003 committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

Aljullu commented on assets/js/blocks/product-collection/inspector-controls/attributes-control.tsx:
```diff
@@ -0,0 +1,81 @@
+import { ProductCollectionQuery } from '../types';
+
+const EDIT_ATTRIBUTES_URL =
+	'/wp-admin/edit.php?post_type=product&page=product_attributes';
```
I think this link won't work if the site is hosted in a subfolder. You can use `ADMIN_URL` constant to prepend the correct path like [here](https://github.com/woocommerce/woocommerce-blocks/blob/ac2b6d6057d2e26b0ac20edba1dcdb5e8530213c/assets/js/blocks/checkout/inner-blocks/checkout-shipping-methods-block/edit.tsx#L118).

imanish003 replied:
That's a good suggestion. Utilizing the ADMIN_URL constant ensures the link works regardless of the site's hosting location in a subfolder. I made the changes in [f9a6813](https://github.com/woocommerce/woocommerce-blocks/pull/9600/commits/f9a6813300b2b9a3aa89f170f7339a4e4efaddc8).

Aljullu commented on assets/js/blocks/product-collection/inspector-controls/attributes-control.tsx:
```diff
@@ -0,0 +1,81 @@
+	woocommerceAttributes,
+	setQueryAttribute,
+}: Props ) => {
+	const [ selected, setSelected ] = useState< { id: number }[] >( [] );
+
+	useEffect( () => {
+		if ( woocommerceAttributes ) {
+			setSelected(
+				woocommerceAttributes.map( ( { termId: id } ) => ( {
+					id,
+				} ) )
+			);
+		}
+	}, [ woocommerceAttributes ] );
```
Do you think we could convert this component to a fully controlled component so it doesn't need to hold its state? Something along these lines:

```suggestion
	const selected = woocommerceAttributes.map( ( { termId: id } ) => ( {
		id,
	} ) );
```

imanish003 replied:
That's a valid idea. By converting the component to a fully controlled component and using the mapped value of woocommerceAttributes for selected, we can eliminate the need for the useState hook. 🤝 

I made the changes in [f9a6813](https://github.com/woocommerce/woocommerce-blocks/pull/9600/commits/f9a6813300b2b9a3aa89f170f7339a4e4efaddc8).

Aljullu commented on assets/js/blocks/product-collection/inspector-controls/index.tsx:
```diff
@@ -50,6 +58,14 @@ const ProductCollectionInspectorControls = (
+						( props.attributes.query
+							.woocommerceAttributes as AttributeMetadata[] )
+					}
+					setQueryAttribute={ setQueryAttributeBind }
```
I'm not sure to understand `setQueryAttributeBind`, is it equivalent to this?

```suggestion
					setQueryAttribute={ ( newQuery ) => {
						setQueryAttribute( props, newQuery );
					} }
```

imanish003 replied:
Yes, You are correct. Do you have any concerns regarding the approach? 🙂

Aljullu replied:
Not really, just a personal preference as I find `bind` a bit difficult to understand. :smile:

Aljullu reviewed (commented):
> @Aljullu, FYI, it seems that @thealexandrelara can't see the `Attributes` filter control. Can you please review and check if you are able to see it? 🙂

It's showing up as expected for me. @thealexandrelara did you check if there are any errors in the console?

@imanish003 I did a quick review and overall it's looking good, but I noticed the `Reset all` button doesn't reset the attributes, I think that's because `woocommerceAttributes` is not defined in `DEFAULT_FILTERS`.

Besides that, I left some minor comments below.

thealexandrelara commented:
> It's showing up as expected for me. @thealexandrelara did you check if there are any errors in the console?

I checked the console and the only error that is showing to me is this one:
![image](https://github.com/woocommerce/woocommerce-blocks/assets/20469356/83eba104-26e6-4c9b-9bce-008af0a1d5ea)

This error was appearing to me in other branches as well, however it was not preventing me from using other blocks. I really don't know what is happening on my side, Maybe I'm going to have to do a fresh install on my testing environment

imanish003 committed:
Address PR feedback & other improvements

1. Added `woocommerceAttributes` to `DEFAULT_FILTERS` in the `product-collection/constants.ts` file to fix `reset all` button issue.

2. Refactored `attributes-control.tsx` to make it more maintainable:
   - The constant `EDIT_ATTRIBUTES_URL` now uses `ADMIN_URL` from `@woocommerce/settings` for a more dynamic URL generation.
   - The interface `Props` has been renamed to `AttributesControlProps` for more explicit naming.
   - Removed the usage of `useState` and `useEffect` for selected attributes. Instead, `selectedAttributes` is now directly derived from `woocommerceAttributes`.
   - The CSS class `woocommerce-product-query-panel__external-link` was renamed to `wc-product-collection-panel__external-link`

3. Deleted the `product-collection/inspector-controls/constants.ts` file which was no longer necessary due to changes in product collection implementation.

These changes contribute to codebase quality, improving readability and maintainability.

imanish003 commented:
Hey @Aljullu,

> @imanish003 I did a quick review and overall it's looking good, but I noticed the Reset all button doesn't reset the attributes, I think that's because woocommerceAttributes is not defined in DEFAULT_FILTERS.

Oh, Yeah. You are right. Good catch 🙌🏻 I have made the improvements/changes in [f9a6813](https://github.com/woocommerce/woocommerce-blocks/pull/9600/commits/f9a6813300b2b9a3aa89f170f7339a4e4efaddc8). Can you please take a look again? 🙂

imanish003 commented:
Hey @thealexandrelara 👋 

> This error was appearing to me in other branches as well, however it was not preventing me from using other blocks. I really don't know what is happening on my side. Maybe I'm going to have to do a fresh install on my testing environment.

Yeah, that error shouldn't be a problem. It makes sense to do a fresh install. Apart from this, I also try the following things in such cases:

Fix 1 - Removing the `build` directory and then running `npm start`. 
Fix 2 - Trying to run `npm install` and `composer install` again.
Fix 3 - Restarting the Local server.
Fix 4 - Making sure that the `wp-config.php` file has `define('JETPACK_AUTOLOAD_DEV', true);`.

imanish003 committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes

Aljullu commented on assets/js/blocks/product-collection/inspector-controls/attributes-control.tsx:
```diff
@@ -0,0 +1,74 @@
+				type={ 'token' }
+			/>
+			<ExternalLink
+				className="wc-product-collection-panel__external-link"
```
This class should have the [`wc-block-editor` prefix](https://github.com/woocommerce/woocommerce-blocks/blob/90760e4be19c5e719d8ae20fcb372d01ce63c191/docs/contributors/coding-guidelines.md?plain=1#LL29C6-L29C21).

Aljullu reviewed (approved):
Thanks for the updates, @imanish003! This is looking great on my end. I left one minor comment, but pre-approving.

imanish003 committed:
Add wc-block-editor prefix to className

imanish003 committed:
Merge branch 'trunk' into 9363-product-collection-filters-attributes
